crates/swc_common/src/plugin.rs

```diff
@@ -8,38 +8,58 @@ use crate::{syntax_pos::Mark, SyntaxContext};
 use anyhow::Error;
 use std::any::type_name;
 
-/// Serialize ast to pass into plugins. This is being called from host side.
-/// TODO: https://github.com/swc-project/swc/issues/3167
-#[cfg(feature = "plugin-base")]
-pub fn serialize_for_plugin<T>(t: &T) -> Result<rkyv::AlignedVec, Error>
+/// A struct wraps internal representation of serialized Program to be passed
+/// across plugin's host to guest. Plugin author should not rely on specific
+/// details serialized byte format struct contains: it is strictly
+/// implementation detail which can change anytime.
+#[cfg(any(feature = "plugin-mode", feature = "plugin-rt"))]
+pub struct SerializedProgram(pub rkyv::AlignedVec);
```

Comment:
I think storing this in `swc_common` is bad for maintenance, because of breaking changes.

Reply:
I made `swc_plugin_js_api` because of this.

Reply:
do you mean swc_common itself's breaking changes, or this struct itself? mind elaborate bit more?

Reply:
I mean a breaking change of this struct is a breaking change of `swc_common` because it's exported.

Reply:
I.. think that's kinda expected? this struct will be only used between host / and the plugin, and by the time we make breaking changes I expect both should have breaking version bump as well.

Reply:
If so, naming seems problematic. We will also serialize other things, not only `swc_ecma_ast::Program`.
How do you think about making the name `Serialized` and renaming `deserialize_for_plugin` as `Serialized.deserialize`, and `serialize_for_plugin` -> `Serialized::new_from`?

Reply:
We need to serialize many things, including diagnostics from a plugin.
Also, in future, we may expand plugin API to bundler/css plugins.

Reply:
Sounds legit to me. Probabaly we need something like `Serialized<T>` then?

Reply:
Oh, seems like a great idea. Yeah, it will prevent deserializing as a different type.

Reply:
actually T seems bit tricky. I'll play with things bit and try to update PR, either having T or not having it.

Reply:
You are right. I took a look at this and it can be hard because it's passed between `extern "C"` boundaries.